annotation.go

```diff
@@ -0,0 +1,386 @@
+	// DeleteStream deletes the stream metadata by name.
+	DeleteStream(ctx context.Context, orgID platform.ID, streamName string) error
+	// DeleteStreamByID deletes the stream metadata by id.
+	DeleteStreamByID(ctx context.Context, orgID, id platform.ID) error
+}
+
+// timeFunc and timeNow are used for running unit tests
+var timeFunc func() time.Time
+
+func timeNow() time.Time {
+	if timeFunc == nil {
+		return time.Now()
+	}
+
+	return timeFunc()
+}
+
+// AnnotationEvent contains fields for annotating an event.
+type AnnotationEvent struct {
+	ID               platform.ID `json:"id,omitempty"` // ID is the annotation ID.
+	AnnotationCreate             // AnnotationCreate defines the common input/output bits of an annotation.
+}
+
+// AnnotationCreate contains user providable fields for annotating an event.
+type AnnotationCreate struct {
+	StreamTag string            `json:"stream,omitempty"`    // StreamTag provides a means to logically group a set of annotated events.
+	Summary   string            `json:"summary"`             // Summary is the only field required to annotate an event.
+	Message   string            `json:"message,omitempty"`   // Message provides more details about the event being annotated.
+	Stickers  map[string]string `json:"stickers,omitempty"`  // Stickers are like tags, but named something obscure to differentiate them from influx tags. They are there to differentiate an annotated event.
+	EndTime   *time.Time        `json:"endTime,omitempty"`   // EndTime is the time of the event being annotated. Defaults to now if not set.
+	StartTime *time.Time        `json:"startTime,omitempty"` // StartTime is the start time of the event being annotated. Defaults to EndTime if not set.
+}
+
+// StoredAnnotation represents annotation data to be stored in the database.
+type StoredAnnotation struct {
+	ID        platform.ID `db:"id"`        // ID is the annotation's id.
+	OrgID     platform.ID `db:"org_id"`    // OrgID is the annotations's owning organization.
+	StreamID  platform.ID `db:"stream_id"` // StreamID is the id of a stream.
+	StreamTag string      `db:"name"`      // StreamTag is the name of a stream (when selecting with join of streams).
+	Summary   string      `db:"summary"`   // Summary is the summary of the annotated event.
+	Message   string      `db:"message"`   // Message is a longer description of the annotated event.
+	Stickers  []string    `db:"stickers"`  // Stickers are additional labels to group annotations by.
+	Duration  string      `db:"duration"`  // Duration is the time range (with zone) of an annotated event.
+	Lower     string      `db:"lower"`     // Lower is the time an annotated event beings.
+	Upper     string      `db:"upper"`     // Upper is the time an annotated event ends.
+}
+
+// Validate validates the creation object.
+func (a *AnnotationCreate) Validate() error {
+	switch s := len([]rune(a.Summary)); {
```

Comment:
The stdlib has `utf8.RuneCountInString` which [this SO post](https://stackoverflow.com/questions/49397913/difference-between-length-of-rune-slice-and-runecountinstring) says is more efficient... not sure if it matters.

I do think that `RuneCountInString` would be less tempting to accidentally "refactor" into an incorrect impl. As-is my first thought was that the `[]rune` wrappers are unnecessary because `len` works on `string`.

Reply:
I updated the code to use `utf8.RuneCountInString`, which I agree makes it more explicit as to what the intent of the code is. This way it will still be consistent with cloud's length counting which takes into account special characters.